# Assignment 10 : ASSOCIATION RULES

In [66]:
#importing important libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [4]:
#load the dataset
online_data = pd.read_excel(r"F:\Data science Course by ExcelR\ASSIGNMENTS\10. Association Rules\Online retail.xlsx")
online_data.head()

shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil
0                             burgers,meatballs,eggs                                                                                                                                                                             
1                                            chutney                                                                                                                                                                             
2                                     turkey,avocado                                                                                                                                                                             
3  mineral water,milk,energy bar,whole wheat rice...                                                                                                                                                                             
4                                     low fat yogurt

We can see this dataset has a long column name so we'll be assigning it as "items".

#### Data Preprocessing:

In [8]:
df = pd.DataFrame(np.array(online_data), columns=['items'])      # pd.DataFrame expects array format data and column names as arguments4
#online_data still has the original data.

In [10]:
df.head()

items
0                             burgers,meatballs,eggs
1                                            chutney
2                                     turkey,avocado
3  mineral water,milk,energy bar,whole wheat rice...
4                                     low fat yogurt

In [12]:
df.shape

(7500, 1)

We have 7500 rows and only one column.

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7500 entries, 0 to 7499
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   items   7500 non-null   object
dtypes: object(1)
memory usage: 58.7+ KB


We have only one column in object type.

Let's check if we have any null values.

In [18]:
df.isnull().sum()

items    0
dtype: int64

Apparently no null values.

In [21]:
df.describe()

items
count      7500
unique     5175
top     cookies
freq        223

From this we can say there are a lot of duplicates in this dataset.

In [24]:
print(df.shape)
df.drop_duplicates(inplace=True)
print(df.shape)

(7500, 1)
(5175, 1)


We had around 2325 duplicate values which are not helpful for our analysis so we removed them.

In [29]:
#Forming transactions : creating meaningful groups if items so find relation
#initiall we need to convert items (data frame format) to 
transactions = df['items'].tolist()                                      #convert all the items tp list
transactions = [transaction.split(',') for transaction in transactions]  #seperate each element in list to make it appear as an atomic item
transactions[0],transactions[1],transactions[2]

(['burgers', 'meatballs', 'eggs'], ['chutney'], ['turkey', 'avocado'])

In [31]:
# !pip install mlxtend       #do this if required

In [33]:
#importing necessary libraries
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules

In [35]:
te = TransactionEncoder()                 # defining TransactionEncoder
te_ary = te.fit_transform(transactions)   # calling the function for transaction data
print(te_ary)                             # showing the data in arrayformat

[[False False False ... False False False]
 [False False False ... False False False]
 [False False False ... False False False]
 ...
 [False False False ... False False False]
 [False False False ... False False False]
 [False False False ... False  True False]]


In [37]:
#Converting the above ndarray format to DataFrame

dataset = pd.DataFrame(te_ary, columns = te.columns_)
dataset

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0          False    False              False      False    False        False   
1          False    False              False      False    False        False   
2          False    False              False      False     True        False   
3          False    False              False      False    False        False   
4          False    False              False      False    False        False   
...          ...      ...                ...        ...      ...          ...   
5170       False    False              False      False    False        False   
5171       False    False              False      False    False        False   
5172       False    False              False      False    False        False   
5173       False    False              False      False    False        False   
5174       False    False              False      False    False        False   

      bacon  barbecue sauce  black tea  blueberries  ...  turkey  \
0     False           False      False        False  ...   False   
1     False           False      False        False  ...   False   
2     False           False      False        False  ...    True   
3     False           False      False        False  ...   False   
4     False           False      False        False  ...   False   
...     ...             ...        ...          ...  ...     ...   
5170  False           False      False        False  ...   False   
5171  False           False      False        False  ...    True   
5172  False           False      False        False  ...   False   
5173  False           False      False        False  ...   False   
5174  False           False      False        False  ...   False   

      vegetables mix  water spray  white wine  whole weat flour  \
0              False        False       False             False   
1              False        False       False             False   
2              False        False       False             False   
3              False        False       False             False   
4              False        False       False             False   
...              ...          ...         ...               ...   
5170           False        False       False             False   
5171           False        False       False             False   
5172           False        False       False             False   
5173           False        False       False             False   
5174           False        False       False             False   

      whole wheat pasta  whole wheat rice   yams  yogurt cake  zucchini  
0                 False             False  False        False     False  
1                 False             False  False        False     False  
2                 False             False  False        False     False  
3                 False              True  False        False     False  
4                 False             False  False        False     False  
...                 ...               ...    ...          ...       ...  
5170              False             False  False        False     False  
5171              False              True  False        False     False  
5172              False             False  False        False     False  
5173              False             False  False        False     False  
5174              False             False  False         True     False  

[5175 rows x 120 columns]

In [39]:
def encode_units(x):
    if x == False:
        return 0
    elif x == True:
        return 1

In [41]:
dataset = dataset.applymap(encode_units)      #mapping True-1 and False-0

C:\Users\peros\AppData\Local\Temp\ipykernel_2356\2862144502.py:1: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  dataset = dataset.applymap(encode_units)      #mapping True-1 and False-0


In [43]:
dataset

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0              0        0                  0          0        0            0   
1              0        0                  0          0        0            0   
2              0        0                  0          0        1            0   
3              0        0                  0          0        0            0   
4              0        0                  0          0        0            0   
...          ...      ...                ...        ...      ...          ...   
5170           0        0                  0          0        0            0   
5171           0        0                  0          0        0            0   
5172           0        0                  0          0        0            0   
5173           0        0                  0          0        0            0   
5174           0        0                  0          0        0            0   

      bacon  barbecue sauce  black tea  blueberries  ...  turkey  \
0         0               0          0            0  ...       0   
1         0               0          0            0  ...       0   
2         0               0          0            0  ...       1   
3         0               0          0            0  ...       0   
4         0               0          0            0  ...       0   
...     ...             ...        ...          ...  ...     ...   
5170      0               0          0            0  ...       0   
5171      0               0          0            0  ...       1   
5172      0               0          0            0  ...       0   
5173      0               0          0            0  ...       0   
5174      0               0          0            0  ...       0   

      vegetables mix  water spray  white wine  whole weat flour  \
0                  0            0           0                 0   
1                  0            0           0                 0   
2                  0            0           0                 0   
3                  0            0           0                 0   
4                  0            0           0                 0   
...              ...          ...         ...               ...   
5170               0            0           0                 0   
5171               0            0           0                 0   
5172               0            0           0                 0   
5173               0            0           0                 0   
5174               0            0           0                 0   

      whole wheat pasta  whole wheat rice  yams  yogurt cake  zucchini  
0                     0                 0     0            0         0  
1                     0                 0     0            0         0  
2                     0                 0     0            0         0  
3                     0                 1     0            0         0  
4                     0                 0     0            0         0  
...                 ...               ...   ...          ...       ...  
5170                  0                 0     0            0         0  
5171                  0                 1     0            0         0  
5172                  0                 0     0            0         0  
5173                  0                 0     0            0         0  
5174                  0                 0     0            1         0  

[5175 rows x 120 columns]

#### Association Rule Mining:

In [286]:
frequent_itemsets = apriori(dataset, min_support=0.1, max_len = 2, use_colnames = True)
rules = association_rules(frequent_itemsets, metric='confidence', min_threshold=0.20, num_itemsets = len(frequent_itemsets))
rules.head(20)

C:\Users\peros\anaconda3\Lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:161: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


Empty DataFrame
Columns: [antecedents, consequents, antecedent support, consequent support, support, confidence, lift, representativity, leverage, conviction, zhangs_metric, jaccard, certainty, kulczynski]
Index: []

This is the structure of the output, unfortunately because of poor selection of min_support and threshod we are not getting any rules as output. Let's fix that.

In [317]:
frequent_itemsets = apriori(dataset, min_support=0.03, max_len = 2, use_colnames = True)
rules = association_rules(frequent_itemsets, metric='confidence', min_threshold=0.3, num_itemsets = len(frequent_itemsets))
rules.head(20)

C:\Users\peros\anaconda3\Lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:161: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


antecedents      consequents  antecedent support  \
0             (burgers)           (eggs)            0.113816   
1             (burgers)  (mineral water)            0.113816   
2                (cake)  (mineral water)            0.103575   
3             (chicken)  (mineral water)            0.083865   
4           (chocolate)  (mineral water)            0.205217   
5                (eggs)  (mineral water)            0.208116   
6   (frozen vegetables)  (mineral water)            0.129855   
7   (frozen vegetables)      (spaghetti)            0.129855   
8         (ground beef)  (mineral water)            0.135845   
9         (ground beef)      (spaghetti)            0.135845   
10     (low fat yogurt)  (mineral water)            0.094879   
11               (milk)  (mineral water)            0.170048   
12          (olive oil)  (mineral water)            0.087536   
13           (pancakes)  (mineral water)            0.125217   
14             (shrimp)  (mineral water)            0.098744   
15               (soup)  (mineral water)            0.070918   
16          (spaghetti)  (mineral water)            0.229565   
17           (tomatoes)  (mineral water)            0.091981   
18          (olive oil)      (spaghetti)            0.087536   
19             (shrimp)      (spaghetti)            0.098744   

    consequent support   support  confidence      lift  representativity  \
0             0.208116  0.036135    0.317487  1.525531               1.0   
1             0.299710  0.034589    0.303905  1.013996               1.0   
2             0.299710  0.037681    0.363806  1.213859               1.0   
3             0.299710  0.032657    0.389401  1.299258               1.0   
4             0.299710  0.073237    0.356874  1.190730               1.0   
5             0.299710  0.070145    0.337047  1.124578               1.0   
6             0.299710  0.050435    0.388393  1.295895               1.0   
7             0.229565  0.039034    0.300595  1.309411               1.0   
8             0.299710  0.058744    0.432432  1.442835               1.0   
9             0.229565  0.055845    0.411095  1.790756               1.0   
10            0.299710  0.032464    0.342159  1.141633               1.0   
11            0.299710  0.067826    0.398864  1.330831               1.0   
12            0.299710  0.038454    0.439294  1.465728               1.0   
13            0.299710  0.048696    0.388889  1.297550               1.0   
14            0.299710  0.033430    0.338552  1.129598               1.0   
15            0.299710  0.033430    0.471390  1.572818               1.0   
16            0.299710  0.085024    0.370370  1.235762               1.0   
17            0.299710  0.034589    0.376050  1.254714               1.0   
18            0.229565  0.032464    0.370861  1.615493               1.0   
19            0.229565  0.030338    0.307241  1.338359               1.0   

    leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
0   0.012448    1.160248       0.388735  0.126437   0.138115    0.245559  
1   0.000477    1.006026       0.015576  0.091280   0.005990    0.209657  
2   0.006639    1.100749       0.196538  0.103066   0.091528    0.244766  
3   0.007522    1.146890       0.251415  0.093062   0.128077    0.249181  
4   0.011731    1.088884       0.201538  0.169651   0.081629    0.300616  
5   0.007770    1.056320       0.139891  0.160265   0.053317    0.285545  
6   0.011516    1.144999       0.262407  0.133028   0.126637    0.278336  
7   0.009224    1.101558       0.271562  0.121834   0.092195    0.235314  
8   0.018030    1.233844       0.355168  0.155897   0.189525    0.314218  
9   0.024660    1.308250       0.510993  0.180400   0.235620    0.327181  
10  0.004028    1.064527       0.137066  0.089648   0.060616    0.225238  
11  0.016861    1.164943       0.299523  0.168750   0.141589    0.312585  
12  0.012219    1.248942       0.348228  0.110249   0.199322    0.283799  
13  0.011167    1.14

#### Analysis and Interpretation:

* **Rule 1 :  burgers --> eggs**
* If a person buys burgers he is also likely to buy eggs.
* Antecedent support of 0.113816 for (burgers) says the item is almost repeated those many times out of all the itemsets taken into consideration.
* Consequent support of 0.208116 for (eggs) says the item is almost repeated those many times out of all the itemsets taken into consideration.
* Since we chose min_support=0.03 all the rules we'll be getting will have support more than 0.03
* For this rule our apriory model gives a confidence score of 0.317487 meaning the model is that sure about the rule found by the model.
* lift greater than 1 indicates those two items have positive dependency on each other.
* leverage and conviction is related to strength and significance of the rule formed by the model.
* Zhang's measure of strength between objects that is burgers and eggs, positive number indicates good combination to buy and negative indicates bad combinations to buy.

* **Similarly we found a total of 20 rules predicted by our model**

* **Rule 2 : burgers --> mineral water**
* **Rule 3 : cake --> mineral water**
* **Rule 4 : cake --> mineral water**
* **Rule 5 : chicken --> mineral water**
* **Rule 6 : chocolate --> mineral water**


* We observe that many of the rules have mineral water as consequent meaning mineral water is being bought with most of the items in the itemset.
* Customers are buying mineral water with almost all of the food items they are purchasing.
* We can also observe that spaghetti is being bought by customers when they buy frozen vegetables or ground beef or olive oil or shrimp, meaning the spaghetti is dependent on those items being bought
* If a customer is buying frozen vegetables, ground beef, olive oil, shrimp they are most likely to buy spaghetti as well.

#### **Interview Questions:**


**1. What is lift and why is it important in Association rules?**
**Answer**
* **Lift** is a measure used in association rule mining to evaluate the strength of a rule by considering the co-occurance of itmes relative to their individual occurances.
* **Formula** : lift = P(A and B)/P(A)*P(B)
* Importance of Lift:
  * **Lift > 1** - Indicates a strong positive association between the items basically they appear more than expected by chance.
  * **Lift = 1** - Suggest No association between the items.
  * **Lift < 1** - Indicates negative association between items
* Lift helps prioritize rules by identifying strong and meaningful relationships.

**2. What is support and Confidence. How do you calculate them?**

**Answer**

**Suppport** is a measure of the frequency of an itemset in the dataset
* **Formula** : support(A) = Number of transaction containing A/Total number of transactions.
* Indicates how often an itemset appears in the data, helping filter rare itemsets.

**Confidence** is the measure of the likelihood of an item B occuring in a transaction given that A is present (A-->B)
* **Formula** : Confidence(A-->B) = Support(A and B)/Support(A)
* Reveals the strength of the rule A-->B, expressing the proportion of transactions with A that also contains B.

**3. What are some limitations or challenges of Association rules mining?**

**Answer**
* **Scalability :** Processing large dataset can be expensive due to exponential growth of candidate itemset.
* **Interpretability :** For complex dataset it is difficult to interpret the generated rules.
* **Redundancy :** Many rules generated may be redundant or trivial, making it difficult to identify the meaningful ones.
* **Threshold Dependency :** Identifying appropriate threshold for support and and confidence may be difficult and subjective to the topic.
* **Rare Itemsets :** Association rules often ignores rare but potentially significant itemsets due to low support values.
* **Data Quality :** Noisy, missing values and inconsistencies in the dataset can impact the quality and accuracy of discovered rules.